Jellyfin.Api/Controllers/UserLibraryController.cs

```diff
@@ -162,7 +179,29 @@ public async Task<ActionResult<QueryResult<BaseItemDto>>> GetIntros([FromRoute,
     [ProducesResponseType(StatusCodes.Status200OK)]
     public ActionResult<UserItemDataDto> MarkFavoriteItem([FromRoute, Required] Guid userId, [FromRoute, Required] Guid itemId)
     {
-        return MarkFavorite(userId, itemId, true);
+        var user = _userManager.GetUserById(userId);
+        if (user is null)
+        {
+            return NotFound();
+        }
+
+        var item = itemId.Equals(default)
+            ? _libraryManager.GetUserRootFolder()
+            : _libraryManager.GetItemById(itemId);
+
+        if (item is null)
+        {
+            return NotFound();
+        }
+
+        if (item is not UserRootFolder
+            // Check the item is visible for the user
+            && !item.IsVisible(user))
+        {
+            return Unauthorized($"{user.Username} is not permitted to access item {item.Name}.");
+        }
```

Comment:
This logic is shared between 4 methods, is there a reasonable way to extract it to a helper method to share code? Perhaps using an out parameter to pass back the return result if access is denied?

Reply:
I played around with it and it doesn't feel like a proper solution because the out parameters (you'd need one for the item and one for the user) can be null. You'd either need additional null checks or explicitly tell the compiler that the variable is not null there which feels kinda dirty.

```c#
    private ActionResult? CheckAccess (Guid userId, Guid itemId, out BaseItem? item, out User? user)
    {
        item = null;
        user = _userManager.GetUserById(userId);
        if (user is null)
        {
            return NotFound();
        }

        item = itemId.Equals(default)
            ? _libraryManager.GetUserRootFolder()
            : _libraryManager.GetItemById(itemId);

        if (item is null)
        {
            return NotFound();
        }

        if (item is not UserRootFolder
            // Check the item is visible for the user
            && !item.IsVisible(user))
        {
            return Unauthorized($"{user.Username} is not permitted to access item {item.Name}.");
        }

        return null;
    }
```

Reply:
Fair enough, I suspected you hadn't done it because it was too unwieldy but I didn't have a dev environment up to play with it and couldn't immediately see why it wouldn't work out.